### DISCUSSION TITLE
How to listen to RTSP endpoint in a browser setting??

### DISCUSSION THREAD

**Discussion by ceyxasm** [bluenviron/mediamtx#3963] [Questions] [answered]
### Question

I have created and published a stream to an RTSP endpoint using mediamtx and ffmpeg.
I am also able to listen to it and capture frames and save them locally via:
`ffmpeg -rtsp_transport tcp -i rtsp://localhost:8554/mystream -frames:v 1 output.jpg`
But when I try to do the same in a browser environment; websocket connection is refused.

Here I am trying to just validate an RTSP EP
```
  async validateRTSPConnection() {
    try {
      const ffmpeg = createFFmpeg({
        corePath: "https://unpkg.com/@ffmpeg/core@0.11.0/dist/ffmpeg-core.js", // Updated core path
        log: true
      });
  
      // Add a timeout for loading
      const loadTimeout = new Promise((_, reject) => 
        setTimeout(() => reject(new Error('FFmpeg load timeout')), 10000)
      );
  
      await Promise.race([ffmpeg.load(), loadTimeout]);
      console.log('Loading successful');
  
      const ffmpegCommand = [
        "-rtsp_transport", "tcp",
        "-i", "rtsp://127.0.0.1:8554/mystream",
        "-t", "5",
        "-f", "null",
        "-"
      ];
  
      await ffmpeg.run(...ffmpegCommand);
      console.log('Command run successful');
      return true;
    } catch (error) {
      console.error('Detailed Error:', {
        message: error.message,
        stack: error.stack,
        name: error.name
      });
      return false;
    }
  }
```
And while the function returns true
![image](https://github.com/user-attachments/assets/2f9c8ace-8886-46f7-9fdb-12e390c83e85)
If I proceed to take snapshots, I get null

**Comment by asturl** [comment]
Unless that FAQ is outdated and you're going by some working example, I don't see how

https://ffmpegwasm.netlify.app/docs/faq/#is-rtsp-supported-by-ffmpegwasm

**Comment by aler9** [comment, marked as answer]
We don't support using WebSocket to pull RTSP streams from the server and we discourage using WASM FFmpeg for reading streams. This stack is difficult to install, power hungry, not robust.

Just use [WebRTC](https://github.com/bluenviron/mediamtx#webrtc) or [HLS](https://github.com/bluenviron/mediamtx#hls) that are protocols meant to be used with browsers. In MediaMTX, conversion between protocols is automatic, so you can publish a stream with RTSP and read it with WebRTC or HLS.

**Reply by cssinate** [reply]
Correct me if I'm wrong - but this is only true if the RTSP codec is one that matches with WebRTC or HLS right? Do you have a recommendation if the RTSP codec is M-JPEG?

**Reply by aler9** [reply]
@cssinate yes. Convert it to H264 by following instructions in the README:
https://github.com/bluenviron/mediamtx?tab=readme-ov-file#remuxing-re-encoding-compression
